### ISSUE TITLE
Run NUnit Tests in Multiple Cultures

### ISSUE THREAD

**Issue by steve09014098** [nunit/nunit#3882]
I can only find reference to a planned improvement in NUnit 3 that I don't think was ever implemented.

I would like to be able to specify an attribute at Assembly, Test Fixture and TestCase level that will automatically duplicate any tests beneath that type in multiple cultures.

So for example, the below code might run the Test Case 3 times, one for each culture specified.

```
[Culture("en-GB, de-DE, zh-CN"]
public class TestClass
{
     [TestCase]
     public void Test()
     {
     }
}
```

It seems odd to me that in 2012 - 2013 this was a hot topic and was stated to be included in the next version of NUnit (3) and I can't find any reference to this since then.

I have a workaround where I can copy out my test code into methods and then set the culture of multiple test cases that call the same method, however retrofitting this to 10,000+ test cases is a major piece of work in and of itself.

Thanks

**Comment by stevenaw** [comment]
I encountered the need for this myself within NUnit's our test suite as I've been trying to add a test to cover https://github.com/nunit/nunit/issues/4246
In that case I would seek to be able to easily run a subset of the whole assembly's tests under multiple cultures. Note that I think this is related but distinct to https://github.com/nunit/nunit/issues/2900 which seeks to run the entire nunit test suite under multiple cultures. One of the concerns raised in that issue was the time commitment to run against a potentially large number of cultures which is something the ability to target only a subset could help with

**Comment by rprouse** [comment]
I like the idea but I think it needs a bit more design first. I think I'd prefer a params array for the cultures rather than one string. How are the tests for each culture named?

**Comment by stevenaw** [comment]
Agreed, I like the idea of params array too so that the culture strings can be kept separate.

For test names, that's the part I've been struggling to wrap my head around design-wise. Most "variables" in the test names naturally map to C# constructs in the form of argument types for methods, etc. But .NET doesn't have a way of "method resolution" by culture which seems to be what we'd want here.

One potentially ugly option fraught with edge cases is to simply suffix it to the end of the method name, before the argument list. A hyphen isn't valid in a C# method name, but I admit I'm rusty on my F# and VB and it's definitely been a while since I've experimented with what's valid in the underlying MSIL itself. This approach would also assume that a culture is always in the full ISO code like so:

```csharp
[Test, Culture("en-US", "en-GB")]
public void Method1() { Assert.Pass(); }

// translates to an test name of 
TestName = `Method1_en-US`
```

however the following is also a perfectly valid way to create a culture in C# and NUnit:
```csharp
[Test, SetCulture("en")]
public void Test2() { Assert.Pass(CultureInfo.CurrentCulture.Name); }
```

And simply suffixing `_en` to the end would inevitably run the risk of a naming conflict with another test method with the literal name `Test2_en`. Perhaps using either `<>` or a backtick as a separator could help here instead of an underscore. I believe the C# compiler uses this when generating type names for anonymous classes or generic types and that these were chosen _because_ they are valid MSIL but not valid in C# or other .NET languages.

```csharp
var c = new { foo = 1 };
Console.WriteLine(c.GetType().Name);   // <>f__AnonymousType0`1

var a = new List<int>();
Console.WriteLine(a.GetType().Name);  // List`1
```

Which _could_ mean something like this as applied here:

```csharp
[Test, SetCulture("en")]
public void Test3() { Assert.Pass(TestContext.CurrentContext.Test.Name); } // outputs "Test3<>Culture=en" ?
```

It looks awkward but would be somewhat safe from naming collisions.

**Comment by stevenaw** [comment]
I rushed completion of that last message, but I'd be very curious to hear other's thoughts on ideas

**Comment by OsirisTerje** [comment]
Although these are valid NUnit test names, they might not be not valid FQNs. They would then go into this group os issues we have from `dotnet test` and Visual Studio. https://github.com/nunit/nunit3-vs-adapter/issues?q=is%3Aopen%20is%3Aissue%20project%3Anunit%2Fnunit3-vs-adapter%2F1 

The runner from Microsoft depends on correctly c# FQNs.  

I haven't looked into using the special characters mentioned above, but it should be easy to check out.  We have to make sure any such changes are tests that actually can be used in filtering expressions.
